- Packages and Modules
- Regular Expressions
- File Handling

- Packages and Modules
    - Package
        - collection of python modules
    - Module
        - a file that contians python code
        - code that define functions,classes and variables
        - These functions can be accessed by using import

In [1]:
# from foldername.pythonfilename import functionname
from package.module import average
average(1,10)

Average:  5.5


In [2]:
from package.mymodule import Even
Even(9)

Odd


In [1]:
from package.module import leapyears
l = int(input("Enter start year: "))
u = int(input("Enter end year: "))
leapyears(l,u)

Enter start year: 1996
Enter end year: 2020
1996 2000 2004 2008 2012 2016 2020 

In [2]:
# Inbuit modules
import math
print(dir(math),end=" ")

['__doc__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc'] 

In [3]:
math.log(10)

2.302585092994046

In [7]:
from math import *
print(pi)
print(factorial(6))
print(floor(6.7))
print(ceil(6.7))

3.141592653589793
720
6
7


In [8]:
import random 
print(dir(random),end=" ")

['BPF', 'LOG4', 'NV_MAGICCONST', 'RECIP_BPF', 'Random', 'SG_MAGICCONST', 'SystemRandom', 'TWOPI', '_BuiltinMethodType', '_MethodType', '_Sequence', '_Set', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_acos', '_bisect', '_ceil', '_cos', '_e', '_exp', '_inst', '_itertools', '_log', '_os', '_pi', '_random', '_sha512', '_sin', '_sqrt', '_test', '_test_generator', '_urandom', '_warn', 'betavariate', 'choice', 'choices', 'expovariate', 'gammavariate', 'gauss', 'getrandbits', 'getstate', 'lognormvariate', 'normalvariate', 'paretovariate', 'randint', 'random', 'randrange', 'sample', 'seed', 'setstate', 'shuffle', 'triangular', 'uniform', 'vonmisesvariate', 'weibullvariate'] 

In [14]:
random.random()

0.526966770842987

In [16]:
random.randint(00000,99999)

5908

In [26]:
random.randrange(0,100,10)

90

In [32]:
random.choice([1,2,3,4,5])

5

In [33]:
random.choices([1,2,3,4,5])

[1]

In [23]:
n =[1,8,9,0,5]
random.shuffle(n)
print(n)

[9, 0, 1, 5, 8]


In [29]:
random.sample(n,k=3)

[9, 0, 8]

In [36]:
random.seed(3)
random.randint(1,100)

31

In [37]:
import time
print(dir(time),end=" ")

['_STRUCT_TM_ITEMS', '__doc__', '__loader__', '__name__', '__package__', '__spec__', 'altzone', 'asctime', 'clock', 'ctime', 'daylight', 'get_clock_info', 'gmtime', 'localtime', 'mktime', 'monotonic', 'monotonic_ns', 'perf_counter', 'perf_counter_ns', 'process_time', 'process_time_ns', 'sleep', 'strftime', 'strptime', 'struct_time', 'thread_time', 'thread_time_ns', 'time', 'time_ns', 'timezone', 'tzname'] 

In [38]:
print(time.localtime())

time.struct_time(tm_year=2020, tm_mon=9, tm_mday=23, tm_hour=8, tm_min=57, tm_sec=45, tm_wday=2, tm_yday=267, tm_isdst=0)


In [39]:
print(time.ctime())

Wed Sep 23 08:58:07 2020


In [41]:
# count number of seconds since epoch(jan1,1970)
print(time.time())

1600831743.0862076


In [73]:
import calendar
print(dir(calendar))

['Calendar', 'EPOCH', 'FRIDAY', 'February', 'HTMLCalendar', 'IllegalMonthError', 'IllegalWeekdayError', 'January', 'LocaleHTMLCalendar', 'LocaleTextCalendar', 'MONDAY', 'SATURDAY', 'SUNDAY', 'THURSDAY', 'TUESDAY', 'TextCalendar', 'WEDNESDAY', '_EPOCH_ORD', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_colwidth', '_locale', '_localized_day', '_localized_month', '_spacing', 'c', 'calendar', 'datetime', 'day_abbr', 'day_name', 'different_locale', 'error', 'firstweekday', 'format', 'formatstring', 'isleap', 'leapdays', 'main', 'mdays', 'month', 'month_abbr', 'month_name', 'monthcalendar', 'monthlen', 'monthrange', 'nextmonth', 'prcal', 'prevmonth', 'prmonth', 'prweek', 'repeat', 'setfirstweekday', 'sys', 'timegm', 'week', 'weekday', 'weekheader']


In [43]:
print(calendar.calendar(2020))

                                  2020

      January                   February                   March
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
       1  2  3  4  5                      1  2                         1
 6  7  8  9 10 11 12       3  4  5  6  7  8  9       2  3  4  5  6  7  8
13 14 15 16 17 18 19      10 11 12 13 14 15 16       9 10 11 12 13 14 15
20 21 22 23 24 25 26      17 18 19 20 21 22 23      16 17 18 19 20 21 22
27 28 29 30 31            24 25 26 27 28 29         23 24 25 26 27 28 29
                                                    30 31

       April                      May                       June
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
       1  2  3  4  5                   1  2  3       1  2  3  4  5  6  7
 6  7  8  9 10 11 12       4  5  6  7  8  9 10       8  9 10 11 12 13 14
13 14 15 16 17 18 19      11 12 13 14 15 16 17      15 16 17 18 19 20 21
20 21 22 23 24 25 26      18 19 20 21 22 

In [45]:
print(calendar.isleap(2021))

False


In [46]:
print(calendar.leapdays(2000,2020))

5


In [48]:
print(calendar.month(2020,6,w=2,l=2))

     June 2020

Mo Tu We Th Fr Sa Su

 1  2  3  4  5  6  7

 8  9 10 11 12 13 14

15 16 17 18 19 20 21

22 23 24 25 26 27 28

29 30




- Regular Expressions
    - sequence of characters to match or to search specific pattern
    - In built package called re
    - 3 Regex Functions
        - match()
            - if re matches at the beginning of the string
        - search()
            - scan through entire string where it matches
        - findall()
            - find all substrings where re matches and returns them in list

In [54]:
import re
re.match('p','ythonp')

In [52]:
import re
re.search('p','python program')

<re.Match object; span=(0, 1), match='p'>

In [51]:
import re
re.findall('p','python program')

['p', 'p']

In [71]:
n = 'my 2 favorite numbers are 19 and 43 0 23'
y = re.findall('[0-9]+',n)
y

['2', '19', '43', '0', '23']

- pattern -- ''/""
- ^ - start
- $ - end
- [0-9] - numbers
- [a-z] - small letters
- [A-z] - Capital letters
- [a-zA-z] - may be both
- [@] - special symbols 
- [0-9]{4} - match 4 digit numbers
- | - means another pattern
- [2468] - single digit multiples of 2
- . - matches any charater except new line
- * - zero or more occurences
- + - one or more occurences

In [60]:
import re
n = 'python online  on one day'
x = re.findall('oni*',n)
print(x)

['on', 'on', 'on', 'on']


In [61]:
import re
n = 'python online  on one day'
x = re.findall('oni+',n)
print(x)

[]


In [ ]:
# phone number validation
# must contain 10 numbers
# start with 6/7/8/9
# + 91
# pattern 
# '^[6-9][0-9]{9}$'
# '^[+][9][1][6-9][0-9]{9}$'

In [65]:
import re
phone = input('Enter your mobile number: ')
pattern = "^[6-9][0-9]{9}$|^[+][9][1][6-9][0-9]{9}$"
if re.match(pattern,phone):
    print("Valid phone number")
else:
    print("Not valid ")

Enter your mobile number: 5678904122
Not valid 


In [ ]:
# Email id validation
# usename@domain.extension
# username specifications
    # length - min -6
               max - 12
    # doesn't start with number/capital letter/special character
    # should start with small letters 
    # special chars - . and _
    # after username should match with @
# domain
    # length - min - 5, max - 9
    # only small letters
# extension
    # length - min - 2.max - 4
    # only small letters

In [70]:
# Password validation
# should start with Capital letter,match @,must have  numbers
# password length 10
# pattern - "^[A-Z][a-z]{5,10}[@][0-9]{3}$"
import re
password = input("Enter your password: ")
pattern = "^[A-Z][a-z]{5,10}[@][0-9]{3}$"
if re.match(pattern,password):
    print("valid")
else:
    print("Not Valid")

Enter your password: apssdc@123
Not Valid


- Function to print reverse numbers of given range in same
    - input -- 1,10
    - output -- 10,9,8,7,6,5,4,3,
- print odd numbers in reverse order
    - input - 123 130
    - output - 129 127 125 123
- with in range(1,10) average of cubes of even numbers

In [72]:
print(dir(list))

['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


In [75]:
l =[1,2,3,4]
for i in l:
    print(i,end=" ")

1 2 3 4 